opt_adapt/opt.py

```diff
@@ -322,60 +644,64 @@ def minimise(
         op.m_progress.append(u)
         op.dJdm_progress.append(dJ)
 
+        # If lr is too small, the difference u-u_ will be 0, and it may cause error
+        if lr < 1e-25:
+            raise fd.ConvergenceError(
+                term_msg + "fail, because control variable didn't move"
+            )
+
         # Check for QoI divergence
         if it > 1 and np.abs(J / np.min(op.J_progress)) > params.dtol:
             raise fd.ConvergenceError(term_msg + "dtol divergence")
 
         # Check for gradient convergence
         if it == 1:
             dJ_init = fd.norm(dJ)
+            J_init = np.abs(J)
         elif fd.norm(dJ) / dJ_init < params.gtol:
             if params.disp > 0:
                 pprint(term_msg + "gtol convergence")
             break
+        # For some situation, convergence should be true, but don't satisfy the above condition
+        elif np.abs(J - op.J_progress[-2]) / J_init < 1e-5:
+            if fd.norm(dJ) / dJ_init < 1e-3:
```

Comment:
It seems odd to check conditions on both objective function and its gradient. Why did you do that?

Reply:
Interesting. I think it would be good to discuss this in our next meeting.